### ISSUE TITLE
volta node installed version not working

### ISSUE THREAD

**Issue by danielpza** [NixOS/nixpkgs#242338]
### Describe the bug

I'm trying to use nodejs/yarn/npm through volta (see https://github.com/NixOS/nixpkgs/pull/217233), with home manager. After setting the VOLTA_HOME variable, running `volta install node` and bit of fiddling trying to make it work I get an error `Volta error: Node is not available.`. Looking into the logs it seems the `libstdc++.so.6` library is not installed at runtime, but not sure if that's the issue

### Steps To Reproduce
1. install `volta` package
2. `volta install node`
3. node --version

### Expected behavior
node --version volta installed node version

### Screenshots
![image](https://github.com/NixOS/nixpkgs/assets/17787042/b6305c65-fcf3-4f98-a5f9-ce9df68e7021)
![image](https://github.com/NixOS/nixpkgs/assets/17787042/f161877f-a46d-477c-82c1-86b0020060d6)
![image](https://github.com/NixOS/nixpkgs/assets/17787042/cf970042-bc53-4571-8645-247fd1ee819b)

### Notify maintainers

package added by @cideM

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
❯ nix-shell -p nix-info --run "nix-info -m"
this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/k57fgpvjv1z5lk7w5p5lhp9ikvj04i0i-nix-info
copying path '/nix/store/k57fgpvjv1z5lk7w5p5lhp9ikvj04i0i-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.37, NixOS, 23.05 (Stoat), 23.05.1784.d8bb6c681cf`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(daniel): `"home-manager, nixpkgs"`
 - channels(root): `"home-manager-23.05.tar.gz, nixos-23.05"`
 - nixpkgs: `/home/daniel/.nix-defexpr/channels/nixpkgs`
```

**Comment by danielpza** [comment]
I also tried using nixos-unstable and installing volta at system level, to no avail:

```nix
  environment.systemPackages = with pkgs; [ volta ];
```

```
❯ echo $VOLTA_HOME
/home/daniel/.volta

❯ volta install node
success: installed and set node@18.16.1 (with npm@9.5.1) as default

❯ which node
/home/daniel/.volta/bin/node

❯ node --version
Volta error: Node is not available.

To run any Node command, first set a default version using `volta install node`
Error details written to /home/daniel/.volta/log/volta-error-2023-07-08_17_26_40.331.log

❯ volta which node
/home/daniel/.volta/tools/image/node/18.16.1/bin/node
 
❯ /home/daniel/.volta/tools/image/node/18.16.1/bin/node --version
bash: /home/daniel/.volta/tools/image/node/18.16.1/bin/node: cannot execute: required file not found
```

**Comment by danielpza** [comment]
was able to fix the issue by using https://github.com/Mic92/nix-ld

**Comment by cideM** [comment]
I'll take a look on my NixOS machine later. Possibly just needs that runtime dep on NixOS, I was only able to test this on my Darwin machine when I submitted it.

**Comment by cideM** [comment]
I can reproduce this on my NixOS machine, obviously. The problem is that Volta downloads the binaries so there's no good way for us to patch them.

I don't know how to fix this but one idea I have is to use `buildFHSEnv` to wrap the shim that Volta creates.

```
❯ tree ~/.volta/ -L 6
/home/fbrs/.volta/
├── bin
│   ├── node -> /nix/store/yrkbqhv7a9m6mbnz9qn1rvfzzm5kr6kp-volta-1.1.1/bin/volta-shim
```

Would it work if we replace the `node` binary here with `buildFHSEnv` so that `~/.volta/bin/node` executes the `volta-shim` in an environment that has `libstdc++.so.6` available?

**Comment by danielpza** [comment]
That looks good to me but I'm also not very experienced with nix.

**Comment by cideM** [comment]
I can't get this to work and I don't know what else I can try.

First of all, the `volta-shim` binary detects if it's called through a symlink or not and if not, it exits with an error. Next, after installing `volta`, when you run `volta install node` it will create symlinks in `~/.volta/bin` so that all binaries (`node`, `npm`...) point at `volta-shim`. This also makes this issue harder to solve, because even if you manage to successfully wrap some command so that the missing library is discoverable at runtime, Volta will link to the original, unwrapped binaries. I assume that this is the reason why they mention in their documentation that all volta binaries need to be in the same folder. So they can link to them relatively to where the `volta` binary lives.

Lastly, I always struggle with this dynamic linking business.

Long story short, I don't think I can fix this issue.

**Comment by danielpza** [comment]
> Long story short, I don't think I can fix this issue. I therefore believe it might be better to mark this as not supported or broken on NixOS

yeah, the nix-ld solution works for me, it was just hard to figure it out. It should be pointed out somewhere, in the package description or perhaps the nixos wiki

**Comment by tim-richter** [comment]
@danielpza Sorry for opening this up again, but what were the steps you needed to take with nix-ld? Currently trying to figure out how to make volta work as well and I can't seem to find a lot of resources on this.

**Comment by danielpza** [comment]
I added these 2 variables to the session environment with home manager:

```nix
    NIX_LD_LIBRARY_PATH =
      lib.makeLibraryPath (with pkgs; [ stdenv.cc.cc openssl ]);
    NIX_LD = lib.fileContents "${pkgs.stdenv.cc}/nix-support/dynamic-linker";
```

I'll try to find a source for this.
